### DISCUSSION TITLE
Nuxt3 useFetch TypeScript

### DISCUSSION THREAD

**Discussion by dev-fredericfox** [nuxt/nuxt#17828] [Questions] [answered]
Using the example from the Nuxt3 Documentation
https://v3.nuxtjs.org/docs/usage/data-fetching

```typescript
<script setup lang="ts">
const { data: mountain } = await useFetch('/api/mountains/everest', { pick: ['title', 'description'] })
</script>

<template>
  <h1>{{ mountain.title }}</h1>
  <p>{{ mountain.description }}</p>
</template>
```

I am quite new to TS, there is probably an easy solution I am not seeing. Is anyone able to explain/link to me in a few words why we get the error `Type 'string' is not assignable to type 'never'.` on { pick: ['title', 'description'] } and how to resolve it?

Thank you!

**Comment by olivervorasai** [comment, marked as answer]
Typescript is saying that the `api` at the address of `/api/mountains/everest` is returning `never`. Basically this means that you don't have an API at the address that Typescript can see. Therefore, you cannot pick out the `title` or the `description` from something that doesn't exist.

What you need to have is a file like this:

```typescript
export default (): { title: string; description: string; height: number } => {
    return {
        title: 'Mt. Everest',
        description: 'Tallest mountain in the world',
        height: 8848,
    };
};
```

in your folder structure like this `server/api/mountains/everest.ts` (https://v3.nuxtjs.org/docs/directory-structure/server)

Also, sometimes typescript can still not see your files correctly depending on the state of your IDE. If you are using VSCode for example, you can do `CTRL + Shift + P` and then type `reload window` to find the reload command.

------------------------------

If instead you are accessing some other url you can help typescript understand what it can choose by defining the type of the return value from the url.

```vue
<script setup lang="ts">
interface Mountain {
    title: string;
    description: string;
}

const { data: mountain } = await useFetch<string, Mountain>(
    'https://www.example.com/mountains',
    {
        pick: ['title', 'description'],
    }
);
</script>
```

**Reply by dev-fredericfox** [reply]
The interface works like a charm, thank you for explaining how to add it to useFetch, I had never seen that specific format.

Quick followup question: in `useFetch<string, Mountain>`, the `<string,` part, before `Mountain>`, does that then refer to the api url itself?

Again, thank you, that was very insightful.

**Reply by olivervorasai** [reply]
I am not sure but looking at the type definition of `useFetch` it would seem that your guess is correct, it does refer to the url.
